File: pci_vf_monitor.py
#!/usr/bin/env python3
"""
OpenNebula PCI probe enhancement: add Physical Function (PF) to Virtual Function (VF) relationship.
This script reads /sys/bus/pci/devices/ and outputs a JSON array of PCI devices,
enriched with PF address for VFs and list of VF addresses for PFs.
"""

import json
import os
import sys

def read_sysfs(path):
    try:
        with open(path, 'r') as f:
            return f.read().strip()
    except (FileNotFoundError, IOError):
        return None

def get_pci_devices():
    devices_path = '/sys/bus/pci/devices'
    devices = {}
    try:
        for dev in os.listdir(devices_path):
            dev_path = os.path.join(devices_path, dev)
            if not os.path.isdir(dev_path):
                continue
            vendor = read_sysfs(os.path.join(dev_path, 'vendor'))
            device = read_sysfs(os.path.join(dev_path, 'device'))
            if not vendor or not device:
                continue
            # Check if it's a VF by looking for physfn link
            physfn_link = os.path.join(dev_path, 'physfn')
            if os.path.islink(physfn_link):
                pf_bdf = os.path.basename(os.readlink(physfn_link))
                role = 'VF'
            else:
                pf_bdf = None
                # Check if it's a PF by looking for virtfn links
                role = 'PF' if os.path.isdir(os.path.join(dev_path, 'virtfn0')) else 'OTHER'
            devices[dev] = {
                'address': dev,
                'vendor': vendor,
                'device': device,
                'role': role,
                'pf_address': pf_bdf,
                'vf_list': [] if role == 'PF' else None
            }
    except OSError as e:
        print(f"Error reading PCI devices: {e}", file=sys.stderr)
        sys.exit(1)
    # Second pass: populate vf_list for PFs
    for dev, info in devices.items():
        if info['role'] == 'PF':
            dev_path = os.path.join(devices_path, dev)
            i = 0
            while True:
                vf_link = os.path.join(dev_path, f'virtfn{i}')
                if os.path.islink(vf_link):
                    vf_bdf = os.path.basename(os.readlink(vf_link))
                    if vf_bdf in devices:
                        info['vf_list'].append(vf_bdf)
                    i += 1
                else:
                    break
    return list(devices.values())

if __name__ == '__main__':
    devices = get_pci_devices()
    print(json.dumps(devices, indent=2))

File: pci_vf_monitor.rb
#!/usr/bin/env ruby
# This module enriches PCI device monitoring with Physical Function (PF) / Virtual Function (VF) relationships.
# It should be integrated into the existing PCI probe in OpenNebula.

require 'json'
require 'fileutils'

module PCIVFMonitor
  SYSFS_PCI_PATH = '/sys/bus/pci/devices'
  NET_CLASS_ID = '0x02'.freeze

  # Parse the existing PCI data array (each element is a hash with :pci_address, :class, etc.)
  # Adds :pf_pci for VFs and optionally an array of VF addresses for PFs.
  def self.enrich_pci_data(pci_devices)
    pci_devices.each do |dev|
      next unless dev[:class] == NET_CLASS_ID
      pci_address = dev[:pci_address]
      pci_path = File.join(SYSFS_PCI_PATH, pci_address)
      next unless Dir.exist?(pci_path)

      # Check if this device is a Virtual Function
      physfn_path = File.join(pci_path, 'physfn')
      if File.symlink?(physfn_path)
        # It's a VF: get PF PCI address from symlink target
        pf_symlink = File.readlink(physfn_path)
        pf_pci = File.basename(pf_symlink)
        dev[:pf_pci] = pf_pci
      else
        # It's a PF (or standalone): check for VFs via sriov_numvfs or numvfs
        sriov_path = File.join(pci_path, 'sriov_numvfs')
        if File.exist?(sriov_path)
          num_vfs = File.read(sriov_path).strip.to_i rescue 0
          if num_vfs > 0
            vfs = []
            # VFs are enumerated as lower addresses with same domain:bus:device.function offsets
            # More robust: read virtfn* symlinks in sysfs
            Dir.glob(File.join(pci_path, 'virtfn*')).sort.each do |vf_link|
              vf_pci = File.basename(File.readlink(vf_link))
              vfs << vf_pci
            end
            dev[:vf_pci_list] = vfs unless vfs.empty?
          end
        end
      end
    end
    pci_devices
  end

  # Convenience method to collect all network devices from sysfs
  def self.collect_all_network_pci
    devices = []
    Dir.foreach(SYSFS_PCI_PATH) do |pci_addr|
      next if pci_addr == '.' || pci_addr == '..'
      pci_path = File.join(SYSFS_PCI_PATH, pci_addr)
      begin
        class_id = File.read(File.join(pci_path, 'class')).strip rescue next
        next unless class_id == NET_CLASS_ID
        devices << {
          pci_address: pci_addr,
          class: class_id
        }
      rescue => e
        # skip inaccessible
      end
    end
    enrich_pci_data(devices)
  end
end

# Example usage (uncomment to test):
# devices = PCIVFMonitor.collect_all_network_pci
# puts JSON.pretty_generate(devices)
